In [1]:
import pandas as pd

In [2]:
# Import the data
file = '../Resources/SFO_Airport_Utility_Consumption.csv'
utilities_df = pd.read_csv(file)

utilities_df.head()

Year  Month Number Month      Utility       Owner   Units       Usage
0  2013             1   Jan   Passengers      Campus     PAX   3209356.0
1  2013             1   Jan          Gas  Commission  Therms    363205.0
2  2013             1   Jan          Gas      Tenant  Therms     49393.0
3  2013             1   Jan  Electricity  Commission     kWh  12904353.0
4  2013             1   Jan  Electricity      Tenant     kWh  14002156.0

# Question: 
### Which utility's usage changed the most from 2013 to 2018?

In [24]:
# Data Collection and Cleaning
# Select only the needed columns
utils_df_cleaning = utilities_df[['Year', 'Utility', 'Units', 'Usage']]
utils_df_cleaning.head()

Year      Utility   Units       Usage
0  2013   Passengers     PAX   3209356.0
1  2013          Gas  Therms    363205.0
2  2013          Gas  Therms     49393.0
3  2013  Electricity     kWh  12904353.0
4  2013  Electricity     kWh  14002156.0

In [25]:
# Scale the Usage column to be more readable
# Rows with "Water" as the utility can be left as is
def scale_to_millions(row):
    if row['Utility']=='Water':
        return row['Usage']/1000000

    return row['Usage']

In [26]:
# Alter the Units column to reflect the changes
def millions_of_units(row):
    if row['Utility']=='Water':
        return row['Units']
    return 'Million' + row['Units']
utils_df_cleaning['Units']=utils_df_cleaning.apply(millions_of_units, axis=1)
print(utils_df_cleaning.head())

   Year      Utility          Units       Usage
0  2013   Passengers     MillionPAX   3209356.0
1  2013          Gas  MillionTherms    363205.0
2  2013          Gas  MillionTherms     49393.0
3  2013  Electricity     MillionkWh  12904353.0
4  2013  Electricity     MillionkWh  14002156.0


/var/folders/0v/k5cxfnrj4_x27cyj9ynqp8qh0000gn/T/ipykernel_20712/381853282.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  utils_df_cleaning['Units']=utils_df_cleaning.apply(millions_of_units, axis=1)


In [28]:
# Combine the Utility and Units columns
# by putting Units in parentheses
def combined_utility_and_units(row):
    return f"{row['Utility']} ({row['Units']})"
utils_df_cleaning['Utility']=utils_df_cleaning.apply(combined_utility_and_units, axis=1)
utils_df_cleaning.head()

/var/folders/0v/k5cxfnrj4_x27cyj9ynqp8qh0000gn/T/ipykernel_20712/1420502252.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  utils_df_cleaning['Utility']=utils_df_cleaning.apply(combined_utility_and_units, axis=1)


Year                   Utility          Units       Usage
0  2013   Passengers (MillionPAX)     MillionPAX   3209356.0
1  2013       Gas (MillionTherms)  MillionTherms    363205.0
2  2013       Gas (MillionTherms)  MillionTherms     49393.0
3  2013  Electricity (MillionkWh)     MillionkWh  12904353.0
4  2013  Electricity (MillionkWh)     MillionkWh  14002156.0

In [30]:
# Create two new DataFrames with data from 2013
# and 2018 that each contain only the Utility and 
# Usage column. Reset the index  for each 
# DataFrame to Utility
utils_2013_df_cleaned = utils_df_cleaning.loc[utils_df_cleaning['Year'] == 2013, ['Utility', 'Usage']].copy().set_index('Utility')
utils_2018_df_cleaned = utils_df_cleaning.loc[utils_df_cleaning['Year'] == 2018, ['Utility', 'Usage']].copy().set_index('Utility')

# Display the first few rows of the DataFrame
print(utils_2013_df_cleaned.head())
print(utils_2018_df_cleaned.head())

                               Usage
Utility                             
Passengers (MillionPAX)    3209356.0
Gas (MillionTherms)         363205.0
Gas (MillionTherms)          49393.0
Electricity (MillionkWh)  12904353.0
Electricity (MillionkWh)  14002156.0
                                 Usage
Utility                               
Passengers (MillionPAX)   4.191545e+06
Gas (MillionTherms)       2.739915e+05
Gas (MillionTherms)       5.766990e+04
Electricity (MillionkWh)  1.369217e+07
Electricity (MillionkWh)  1.234269e+07


In [38]:
# Analyze
utlities=utils_2013_df_cleaned.index.unique()
totals_2013={'Period':2013}
totals_2018={'Period':2018}
# Calculate the totals for each utility
for utility in utlities:
    totals_2013['Utility']=utils_2013_df_cleaned.loc[utility, 'Usage'].sum()
    totals_2018['Utility']=utils_2018_df_cleaned.loc[utility, 'Usage'].sum()
utlities 
print(totals_2013)
print(totals_2018)

{'Period': 2013, 'Utility': 390.932476}
{'Period': 2018, 'Utility': 410.236112}


In [ ]:
def get_percentage(original, final):
    return round((final-original)/original,3)*100

In [9]:
# Calculate the change per utility as a percentage
# of each utility's 2013 total.





# Create a dataframe with the results



Utility   2013   2018  Difference  Change %
0   Passengers (Million PAX)   45.0   57.8        12.8      28.4
1       Gas (Million Therms)    3.9    3.6        -0.3      -6.8
2  Electricity (Million kWh)  330.3  311.2       -19.1      -5.8
3    Water (Million Gallons)  390.9  410.2        19.3       4.9

In [10]:
# Set the index to the utility column

# Sort the rows based on Change %


2013   2018  Difference  Change %
Utility                                                      
Passengers (Million PAX)    45.0   57.8        12.8      28.4
Water (Million Gallons)    390.9  410.2        19.3       4.9
Electricity (Million kWh)  330.3  311.2       -19.1      -5.8
Gas (Million Therms)         3.9    3.6        -0.3      -6.8

# Summary
The number of passengers using the airport grew 28% in just 5 short years! Water consumption rose slightly at 4.9%, but despite the increase in airport traffic, Electricity and Gas usage were both down in 2018 compared to 2013, with Gas leading the charge with a 6.8% decline. 